Allows to use tag or tags in aws_autoscaling_group resource - fixes #101

ezraroi opened the pull request:
The suggested solution is to detect if the user used `tags` or `tag` and based on this to decide if we add other `tag` blocks or we used `concat` method to append the tags that terratag wants to add to the current ones

ezraroi committed:
fixes #101

yaronya commented:
Hi @ezraroi 
Thanks for your contribution! 🙏 

Please make sure:
1. PR has a valid title and description (we should be adding PR guidelines soon, sorry bout that tho)
2. CI tests pass

ezraroi commented:
fixes #101

ezraroi commented:
@yaronya looks like it took it 2 days to run the tests

ezraroi committed:
fixing logic

ezraroi committed:
fixed new test

ezraroi committed:
fixed tests according:
https://github.com/terraform-google-modules/terraform-google-cloud-storage/issues/25

ezraroi commented:
Also removed the `bucket_policy_only` from tests as it deprecated

yaronya commented:
> @yaronya looks like it took it 2 days to run the tests

@ezraroi yup we required approval for runs of forked repositories of first-time contributors 🙏 
Sorry for the delay - I've now approved your recent changes and it tests seem to pass

eranelbaz reviewed (changes requested):
Hey!
from first impression looks like `ConcatTagsToTagsBlock` is copy-paste from `TagBlock`,
and its also hard to read and reason with this pr without any description that explain what you did.

I think what we actually want to do here is make `TagBlock` support your issue

ezraroi commented on tagging/tagging.go:
```diff
@@ -47,6 +47,22 @@ func TagBlock(args TagBlockArgs) string {
 	return newTagsValue
 }
 
+func ConcatTagsToTagsBlock(args TagBlockArgs) string {
```
No to Tag, but to tags. If the user used the Tags block we need to append the tags to the Tags block. Otherwise, we can just add more tag blocks

eranelbaz replied:
So `TagToTags` since you set `TagsToTags` its hard to understand what you append where

ezraroi commented:
> 

The issue that is linked to the PR is explaining the problem we face

ezraroi commented:
> 

Also it is not copy-paste, it is not doing the same thing.
I dont need to call `convert.MoveExistingTags` in my code and also there is no need to search if there is a block as this is a concat function that adds tags to existing block.

I will try to extract some common code from both of them

ezraroi committed:
fixed code review

eranelbaz commented:
> The issue that is linked to the PR is explaining the problem we face

The issue explain the problem, the PR description explain the solution for the issue

> I will try to extract some common code from both of them

Thanks 🙏

ezraroi committed:
fixed code review

eranelbaz reviewed (changes requested):
> I dont need to call `convert.MoveExistingTags` in my code and also there is no need to search if there is a block as this is a > concat function that adds tags to existing block.

Looks like you are calling that here 

https://github.com/ezraroi/terratag/blob/d4a263840842d66218d40866802ceb1640704e76/tagging/tagging.go#L47-L50

The current architecture of this code should be in a different way - 
for each specific logic of a resource we run the relevant logic inside the cloud-provider file, in this case - `tagging/aws.go`
then we can change the way this resource is written by covert the `tags` block to multiple `tag` blocks then the rest of our code can run at the same way

you can take a look at the current comment we have 
https://github.com/env0/terratag/blob/740166c4aba6b792fe43635c0464f99cd96a49a0/tagging/aws.go#L8-L10

can we do that?

Thanks

ezraroi commented:
> https://github.com/ezraroi/terratag/blob/d4a263840842d66218d40866802ceb1640704e76/tagging/tagging.go#L47-L50

Sorry, yes i am calling it.
I extracted some of the common code to share methods anyway.

So you suggest to change the `tags` block to multiple `tag` blocks? Why? this will change the code the user provided in his terraform file. I think the current implementation does not break what you wrote.

The comment you mentioned is the problem that i am trying to fix. The idea is to support both use cases.

ezraroi committed:
Merge branch 'master' into git_issue_101

eranelbaz commented:
> > [ezraroi/terratag@`d4a2638`/tagging/tagging.go#L47-L50](https://github.com/ezraroi/terratag/blob/d4a263840842d66218d40866802ceb1640704e76/tagging/tagging.go#L47-L50)
> 
> Sorry, yes i am calling it. I extracted some of the common code to share methods anyway.
> 
> So you suggest to change the `tags` block to multiple `tag` blocks? Why? this will change the code the user provided in his terraform file. I think the current implementation does not break what you wrote.
> 
> The comment you mentioned is the problem that i am trying to fix. The idea is to support both use cases.

So the way terratag works today is by changing the `tf` code of the client in order to add those `tag`/`tags` attributes for the relevant resources

I don't find any reasons why we shouldn't do the same here

avnerenv0 commented:
Hey @ezraroi 

Please respond to @eranelbaz 's requests.
If this PR won't be touched in 3 days I will close it for being stale.

Thank the contribution.

avnerenv0 commented:
Closing stale

jeffliu527 commented:
met with the same issue while creating an AWS bastion server using: https://github.com/Guimove/terraform-aws-bastion , if @ezraroj 's solution cant be accepted, pls advise how to fix this issue.

Error message:

│ Error: Conflicting configuration arguments
│
│ with module.bastion.module.bastion.aws_autoscaling_group.bastion_auto_scaling_group,
│ on .terraform/modules/bastion.bastion/main.terratag.tf line 330, in resource "aws_autoscaling_group" "bastion_auto_scaling_group":
│ 330: tags = concat(
│ 331: tolist([tomap({
│ 332: "key" = "Name"
│ 333: "value" = "ASG-${local.name_prefix}"
│ 334: "propagate_at_launch" = true
│ 335: })]),
│ 336: local.tags_asg_format
│ 337: )
│
│ "tags": conflicts with tag
